libraries/utils/src/render-lines-on-sp.ts

```diff
@@ -63,77 +61,69 @@ export class ScatterPlotLinesRenderer {
-                        lineOpts.id === this.currentLineIdx ? 3 : 1;
+                    const color = this.lines.colors?.[i] ? this.lines.colors?.[i] : '0,128,0';
+                    const opacity = this.lines.opacities?.[i] ? this.lines.opacities?.[i] : 1;
+                    this.ctx.strokeStyle = `rgba(${color},${opacity})`;
+                    this.ctx.lineWidth = this.lines.widths?.[i] ? i === this.currentLineIdx ? this.lines.widths?.[i] + 2 :
+                    this.lines.widths?.[i] : i === this.currentLineIdx ? 3 : 1;
                     let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
                     if (this.numberOfLinesPerPair[pairStr] > 1 || linesPerPairCounter[pairStr] > 1) {
                         midPoint = this.findPerpendicularPointOnCurve(linesPerPairCounter[pairStr],
                             pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
                     }
                     line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
-                    if (lineOpts.directionCol) {
-                        const arrowPoint = this.getArrowPoint(lineOpts, lineOpts.points[0], lineOpts.points[1], pointFrom, pointTo);
-                        if (arrowPoint)
-                            this.canvasArrow(line, arrowPoint.x, arrowPoint.y, midPoint.x, midPoint.y);
-                    }
-                    ctx.beginPath();
-                    ctx.stroke(line);
-                    ctx.closePath();
+                    if (this.lines.drawArrows?.getBit(i))
+                        this.canvasArrow(line, bX, bY, midPoint.x, midPoint.y);
+                    this.ctx.beginPath();
+                    this.ctx.stroke(line);
+                    this.ctx.closePath();
                 }
             }
         }
-
     }
 
-    countLinesPerPairAndSort(): void {
-        for (let i = 0; i < this.lines.length; i++) {
-            const pair = this.lines[i].points;
-            const p1 = pair[0] > pair[1] ? pair[1] : pair[0];
-            const p2 = pair[0] > pair[1] ? pair[0] : pair[1];
-            pair[0] = p1;
-            pair[1] = p2;
-            const pairStr = `${this.lines[i].points[0]}#${this.lines[i].points[1]}`;
-            if (!this.numberOfLinesPerPair[pairStr])
-                this.numberOfLinesPerPair[pairStr] = 1;
-            else
-                this.numberOfLinesPerPair[pairStr]++;
+    countLinesPerPair(): void {
+        for (let i = 0; i < this.lines.from.length; i++) {
+            const pairStr = this.getStringPair(this.lines.from[i], this.lines.to[i]);
+            !this.numberOfLinesPerPair[pairStr] ? this.numberOfLinesPerPair[pairStr] = 1 : this.numberOfLinesPerPair[pairStr]++;
         }
     }
 
     getStringPair(point1: number, point2: number): string {
-        return `${point1}#${point2}`;
+        let p1 = point1;
+        let p2 = point2;
+        if (point1 > point2) {
+            p1 = point2;
+            p2 = point1;
+        }
+        return `${p1}#${p2}`;
```

Comment:
Very inefficient. Strings are expensive.

Reply:
please see the comment above